File: python/use_cases/completion/completion_proxy_client.py
import argparse
import asyncio
import sys
import openai
from openai.types import Completion
from openai.types.chat import ChatCompletion
from decouple import config

openai_api_key = config("OPENAI_API_KEY")
helvia_rag_cache_token = config("RAG_BUDDY_TOKEN")
base_url = f"{config('PROXY_HOST', default='', cast=str)}/proxy/sem/{config('OPENAI_VERSION', default='', cast=str)}"


async def test_completion_async_stream(prompt: str, cache_control: str):
    # Pass the cache control header, if provided
    cache_control_header = {}
    if cache_control:
        cache_control_header = {
            "Helvia-RAG-Buddy-Cache-Control": ", ".join(cache_control)
        }

    headers = {"Helvia-RAG-Buddy-Token": helvia_rag_cache_token, **cache_control_header}

    client = openai.AsyncOpenAI(
        api_key=openai_api_key,
        base_url=base_url,
        timeout=10,
        default_headers=headers,
    )


    raw_response: ChatCompletion = await client.chat.completions.with_raw_response.create(
        prompt=prompt,
        model="gpt-4o",
        stream=True,
    )

    stream = raw_response.parse()

    # Assert response body
    assert isinstance(stream, openai.AsyncStream)
    
    collected_chunks = []
    completion_text = ""
    # iterate through the stream, if it breaks, the test failed
    print("\n")
    async for chunk in stream:
        collected_chunks.append(chunk)
        finish_reason = chunk.choices[0].finish_reason
        if finish_reason is not None:
            break
        chunk_text = chunk.choices[0].text
        print(chunk_text)
        completion_text += chunk_text
    print("\n")
    print(raw_response.headers)

    if "Helvia-RAG-Buddy-Cache-Status" in raw_response.headers:
        cache_status = raw_response.headers["Helvia-RAG-Buddy-Cache-Status"]
        cache_hit = int(cache_status) if cache_status not in ["None", None] else None
        print(f"Cache hit: {cache_hit}")

async def test_completion_async(prompt: str, cache_control: str):
    # Pass the cache control header, if provided
    cache_control_header = {}
    if cache_control:
        cache_control_header = {
            "Helvia-RAG-Buddy-Cache-Control": ", ".join(cache_control)
        }

    headers = {
        "Helvia-RAG-Buddy-Token": helvia_rag_cache_token,
        **cache_control_header,
    }

    client = openai.AsyncOpenAI(
        api_key=openai_api_key,
        base_url=base_url,
        timeout=10,
        default_headers=headers,
    )

    raw_response = await client.chat.completions.with_raw_response.create(
        messages=[            
            {"role": "user", "content": prompt
             },
        ],
        model="gpt-4o",
        stream=False
    )

    completion = raw_response.parse()

    # Assert response body
    assert isinstance(completion, ChatCompletion)
    assert len(completion.choices) > 0
    assert completion.choices[0].message
    print(completion.choices[0].message)
    print(completion.model_dump_json(indent=2))
    print(raw_response.headers)

    if "Helvia-RAG-Buddy-Cache-Status" in raw_response.headers:
        cache_status = raw_response.headers["Helvia-RAG-Buddy-Cache-Status"]
        cache_hit = int(cache_status) if cache_status not in ["None", None] else None
        print(f"Cache hit: {cache_hit}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Process a prompt for completion.")

    parser.add_argument(
        "--stream", action="store_true", help="Enable streaming completion"
    )
    parser.add_argument(
        "--cache-control",
        choices=["no-cache", "no-store"],
        nargs="*",
        default=[],
        help="Set SemCache cache-control header",
    )

    args = parser.parse_args()

    # After parsing arguments:
    print("Please enter the prompt you want to complete:")

    # Read prompt from standard input
    prompt = sys.stdin.readline().rstrip()

    if args.stream:
        asyncio.run(test_completion_async_stream(prompt, args.cache_control))
    else:
        asyncio.run(test_completion_async(prompt, args.cache_control))

File: typescript/use_cases/completion/completionProxyClient.ts
import * as path from "path";
import * as dotenv from "dotenv";
import * as readline from "readline";
import { Command, Option } from "commander";
import OpenAI from "openai";
import { Completion } from "openai/resources/completions";

interface IProgramOptions {
  stream: boolean;
  cacheControl: ("no-cache" | "no-store")[];
}

const envFilePath = path.resolve(__dirname, "../../.env.sample");
dotenv.config({ path: envFilePath });

const OPENAI_API_KEY = process.env.OPENAI_API_KEY;
const PROXY_HOST = process.env.PROXY_HOST;
const RAG_BUDDY_TOKEN = process.env.RAG_BUDDY_TOKEN;
const OPENAI_VERSION = process.env.OPENAI_VERSION;

const baseURL = `${PROXY_HOST}/proxy/sem/${OPENAI_VERSION}`;

async function testCompletionAsyncStream(
  prompt: string,
  cacheControl: string[]
) {
  // Pass the cache control header, if provided
  let cacheControlHeader = {};
  if (cacheControl.length) {
    cacheControlHeader = {
      "Helvia-RAG-Buddy-Cache-Control": cacheControl.join(", "),
    };
  }

  const headers = {
    "Helvia-RAG-Buddy-Token": RAG_BUDDY_TOKEN,
    ...cacheControlHeader,
  };

  const client = new OpenAI({
    apiKey: OPENAI_API_KEY,
    baseURL,
    defaultHeaders: headers,
    timeout: 10000,
  });

  const { data: stream, response } = await client.completions
    .create({
      prompt,
      model: "gpt-3.5-turbo-instruct",
      stream: true,
    })
    .withResponse();

  const chunks: Completion[] = [];
  let completionText = "";
  console.log("\n");
  // Iterate through the stream, if it breaks, the test failed
  for await (const chunk of stream) {
    chunks.push(chunk);
    if (chunk.choices[0].finish_reason) {
      break;
    }
    const chunkText = chunk.choices[0].text || "";
    if (chunkText) {
      console.log(chunkText);
      completionText += chunkText;
    }
  }

  console.log("\n", response.headers);

  if (response.headers.has("Helvia-RAG-Buddy-Cache-Status")) {
    const status = response.headers.get("Helvia-RAG-Buddy-Cache-Status");
    const cacheHit = !!(status && status !== "null");
    console.log(`\nCache hit: ${cacheHit}`);
  }
}

async function testCompletionAsync(prompt: string, cacheControl: string[]) {
  // Pass the cache control header, if provided
  let cacheControlHeader = {};
  if (cacheControl.length) {
    cacheControlHeader = {
      "Helvia-RAG-Buddy-Cache-Control": cacheControl.join(", "),
    };
  }

  const headers = {
    "Helvia-RAG-Buddy-Token": RAG_BUDDY_TOKEN,
    ...cacheControlHeader,
  };

  const client = new OpenAI({
    apiKey: OPENAI_API_KEY,
    baseURL,
    defaultHeaders: headers,
    timeout: 10000,
  });

  const response = await client.completions
    .create({
      prompt,
      model: "gpt-3.5-turbo-instruct",
    })
    .asResponse();

  const data = await response.json();

  console.log("\n", data.choices[0].text);
  console.log("\n", "Model: ", data.model);
  console.log("\n", "Headers: ", response.headers);

  if (response.headers.has("Helvia-RAG-Buddy-Cache-Status")) {
    const status = response.headers.get("Helvia-RAG-Buddy-Cache-Status");
    const cacheHit = !!(status && status !== "null");
    console.log(`\nCache hit: ${cacheHit}`);
  }
}

function main() {
  const program = new Command();

  program
    .description("Process a prompt for completion.")
    .option("--stream", "Enable streaming completion", false)
    .addOption(
      new Option(
        "--cache-control <values...>",
        "Set SemCache cache-control header"
      )
        .default([])
        .choices(["no-cache", "no-store"])
    )
    .parse(process.argv);

  const options: IProgramOptions = program.opts();

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  rl.question("Please enter the prompt you want to complete: ", (prompt) => {
    rl.close();
    options.stream
      ? testCompletionAsyncStream(prompt, options.cacheControl)
      : testCompletionAsync(prompt, options.cacheControl);
  });
}

main();
